overlord/servicestate/quota_control.go

```diff
@@ -58,7 +67,59 @@ func EnsureQuotaUsability() (restore func()) {
 	}
 }
 
+func checkMemoryCGroupEnabled() {
+	memoryCGroupError = memoryCGroupEnabled()
+}
+
+// added to enable path update for testing purposes
+func SetCGroupsFilePath(path string) {
+	CGroupsFilePath = path
+	checkMemoryCGroupEnabled()
+}
+
+// since the control groups can be enabled/disabled without the kernel config the only
+// way to identify the status of memory control groups is via /proc/cgroups
+// "cat /proc/cgroups | grep memory" returns the active status of memory control group
+// and the 3rd parameter is the status
+// 0 => false => disabled
+// 1 => true => enabled
+func memoryCGroupEnabled() error {
+	cgroupsFile, err := os.Open(CGroupsFilePath)
+	if err != nil {
+		return fmt.Errorf("cannot open cgroups file: %v", err)
+	}
+	defer cgroupsFile.Close()
+	scanner := bufio.NewScanner(cgroupsFile)
+	scanner.Split(bufio.ScanLines)
+	for scanner.Scan() {
+		line := scanner.Text()
+		if strings.HasPrefix(line, "memory\t") {
+			memoryCgroupValues := strings.Fields(line)
+			if len(memoryCgroupValues) >= 4 { // assuming any increase in size will lead to values added to the end
+				isMemoryEnabled := memoryCgroupValues[3] == "1"
+				if !isMemoryEnabled {
+					return fmt.Errorf("cannot retrieve quota information, memory cgroup is disabled on this system")
```

Comment:
This function does not know that the caller wants to retrieve the quota information: here we should just return "memory cgroup is disabled on this system" and let the caller enrich the error message with more context.

Reply:
The "cannot" is a tendency, as you wrote, not a rule. We should write "cannot" instead of "failed to", "couldn't", etc., but it does not mean that all errors need to start with it.